man/ioreplay.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.38.2.
.TH IOREPLAY "1" "August 2012" "ioreplay v1.3, part of IOapps" "User Commands"
.SH NAME
ioreplay \- IO traces replayer
.SH SYNOPSIS
.B ioreplay
[\fIOPERATION\fR] \fI-f <file> \fR[\fIOPTIONS\fR]
.SH DESCRIPTION
Replicates all IO syscalls defined in file by \fB\-f\fR option.
.PP
ioreplay is primary used to replicate recorded IO system calls.
In order to do that, several other helper functionality exists.
.PP
Usage: ioreplay \fB\-c\fR \fB\-f\fR <file> [\-F <format>] [\-o <out>] [\-v]
.IP
converts <file> in format <format> to binary form into file <out>
.PP
Usage: ioreplay \fB\-S\fR \fB\-f\fR <file> [\-v]
.IP
displays some statistics about syscalls recorded in <file> (must be in strace format)
.PP
Usage: ioreplay \fB\-P\fR \fB\-f\fR <file> [\-F <format>] [\-v]
.IP
prints syscalls in normalized format
.PP
Usage: ioreplay \fB\-C\fR \fB\-f\fR <file> [\-F <format>] [\-i <file>] [\-m <file>] [\-v]
.IP
checks whether local enviroment is ready for replaying traces recorded in <file>.
.PP
Usage: ioreplay \fB\-r\fR \fB\-f\fR <file> [\-F <format>] [\-t <mode>] [\-s <factor>] [\-b <number>] [\-i <file>] [\-m <file>] [\-v]
.IP
replicates traces recorded in <file>. Use \fB\-C\fR prior to running this.
.TP
\fB\-b\fR \fB\-\-bind\fR <number>
bind replicating process to processor number <number>
.TP
\fB\-c\fR \fB\-\-convert\fR
file to binary form, see also \fB\-o\fR
.TP
\fB\-C\fR \fB\-\-check\fR
checks that all operations recorded in the file specied by \fB\-f\fR will
succeed (ie. will result in same return code).
It takes \fB\-i\fR and \fB\-m\fR into account. See also \fB\-p\fR.
.TP
\fB\-d\fR \fB\-\-dont\-fix\fR
turns off fixing of missing system calls (uncomplete strace output support)
.TP
\fB\-f\fR \fB\-\-file\fR <file>
sets filename to <file>
.TP
\fB\-F\fR \fB\-\-format\fR <fmt>
specifies input format of the file.
Options: strace, bin.
Check README for details. Default is strace.
.TP
\fB\-h\fR \fB\-\-help\fR
prints this message
.TP
\fB\-i\fR \fB\-\-ignore\fR <file>
sets file containing names which we should not touch during
replaying. I.e. no syscall operation will be performed on given file.
.TP
\fB\-m\fR \fB\-\-map\fR <file>
sets containing file names mapping. When opening file,
if there is mapping for it, it will open mapped file instead.
See README for more information.
.TP
\fB\-o\fR \fB\-\-output\fR <file>
output filename when converting. Default: strace.bin.
.TP
\fB\-p\fR \fB\-\-prepare\fR
will prepare all files accesses recorded in file specified by \fB\-f\fR,
so every IO operation will return with same exit code as in original
application. See also \fB\-i\fR and/or \fB\-m\fR parameters.
Do nothing at the moment.
.TP
\fB\-P\fR \fB\-\-print\fR
prints recorded syscalls in normalized format regardless the format
in which are the syscalls stored now
.TP
\fB\-r\fR \fB\-\-replicate\fR
will replicate every operation stored in file specified by \fB\-f\fR
.HP
\fB\-s\fR \fB\-\-scale\fR <factor> scales delays between calls by the factor <factor>. Used with \fB\-r\fR.
.TP
\fB\-S\fR \fB\-\-stats\fR
generate stats when processing the file. Can be combined with other
options.
.TP
\fB\-t\fR \fB\-\-timing\fR
sets timing mode for replication. Options available:
diff  \- default mode. makes sure that gaps between calls are the same as in the original run.
asap  \- makes calls one just after another.
exact \- makes sure that calls are (approximately) done in the same time as in the original run
.IP
(relative from start of the application)
.HP
\fB\-v\fR \fB\-\-verbose\fR be more verbose (do nothing at the moment)
.HP
\fB\-V\fR \fB\-\-version\fR prints version and exits.
.SH AUTHOR
Written by Jiri Horky, 2010
.SH COPYRIGHT
Copyright \(co 2010 Jiri Horky <jiri.horky@gmail.com>
.PP
License GPLv2: GNU GPL version 2 <http://gnu.org/licenses/gpl.html>
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
.BR ioprofiler(1)


CHANGELOG
v1.4-r1
	* Fix: tarball included .svn files
	* Fix: ioprofiler.1 manpage is now included in install
	* Fix: Makefiles cleanup
	* No new features

 -- Jiri Horky <jiri.horky@gmail.com>  Fri, 09 Apr 2013 20:20:05 +0200

v1.4
	* New feature: ioprofiler can open traces from a file specified on the command line as an argument
	* New feature: Help menu for ioprofiler to aid people understand what it is useful for and to point to wiki
	* New feature: ioprofiler-trace wrapper around strace to help getting traces
	* Bug fix: Histograms now show 99% of the data - scale is automatically adjusted
	* Bug fix: "Show stats" button in ioprofiler works for histograms as well
	* Some minor bug fixes

 -- Jiri Horky <jiri.horky@gmail.com>  Fri, 05 Apr 2013 14:20:05 +0200

v1.3
   * New feature: Ioprofiler can now read incomplete traces
   * New feature: Processes' info is now kept in hash tables to improve performance
	* Bug fix: various bugs fixed

 -- Jiri Horky <jiri.horky@gmail.com>  Wed, 21 Sep 2011 08:59:05 +0200

v1.2
   * New feature: support for sendfile syscall
   * New feature: support for incomplete traces (obtained by strace -p) for ioreplay

 -- Jiri Horky <jiri.horky@gmail.com>  Sun, 10 Apr 2011 11:36:22 +0100

v1.1
   * Bug fixed: ioprofiler can now re-open files without restarting the application
   * Bug fixed: some memory leaks and minor bug fixed
   * New feature: support for pread() and pwrite() system calls added

 -- Jiri Horky <jiri.horky@gmail.com>  Thu, 02 Jan 2011 12:56:30 +0100

v1.0
   * Initial release

 -- Jiri Horky <jiri.horky@gmail.com>  Thu, 02 Nov 2010 20:50:01 +0100



trunk/ioreplay.c
/* IOapps, IO profiler and IO traces replayer

    Copyright (C) 2010 Jiri Horky <jiri.horky@gmail.com>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <fcntl.h>
#include <libgen.h>
#include "common.h"
#include "ioreplay.h"
#include "print.h"
#include "replicate.h"
#include "stats.h"
#include "simulate.h"
#include "in_strace.h"
#include "in_binary.h"

#define FORMAT_STRACE "strace"
#define FORMAT_BIN "bin"

static struct option ioreplay_options[] = {
   /* name        has_arg flag  value */
   { "bind",			1,		NULL,	'b' },
   { "convert",		0,		NULL,	'c' },
   { "check",			0,		NULL,	'C' },
   { "dont-fix",		0,		NULL,	'd' },
   { "file",			1,		NULL,	'f' },
   { "format",			1,		NULL,	'F' },
   { "help",			0,		NULL,	'h' },
   { "ignore",			1,		NULL,	'i' },
   { "map",				0,		NULL,	'm' },
   { "output",			1,		NULL,	'o' },
   { "replicate",		0,		NULL,	'r' },
   { "prepare",		0,		NULL,	'p' },
   { "print",		0,		NULL,	'P' },
   { "scale",			1,		NULL,	's' },
   { "stats",			1,		NULL,	'S' },
   { "timing",			1,		NULL,	't' },
   { "verbose",		0,		NULL,	'v' },
   { "version",		0,		NULL,	'V' },
   { NULL,				0,		NULL,	0 }
};

struct timeval global_start;

void help(char * name) {
	printf("Replicates all IO syscalls defined in file by -f option.\n\n");
printf("Usage: %s [OPERATION] -f <file> [OPTIONS]\n\n", name);
printf("%s is primary used to replicate recorded IO system calls.\n\
In order to do that, several other helper functionality exists.\n\n", name);

printf("Usage: %s -c -f <file> [-F <format>] [-o <out>] [-v]\n", name);
printf("   converts <file> in format <format> to binary form into file <out>\n\n");
printf("Usage: %s -S -f <file> [-v]\n", name);
printf("   displays some statistics about syscalls recorded in <file> (must be in " FORMAT_STRACE " format)\n\n");
printf("Usage: %s -P -f <file> [-F <format>] [-v]\n", name);
printf("   prints syscalls in normalized format\n\n");
printf("Usage: %s -C -f <file> [-F <format>] [-i <file>] [-m <file>] [-v]\n", name);
printf("   checks whether local enviroment is ready for replaying traces recorded in <file>.\n\n");
printf("Usage: %s -r -f <file> [-F <format>] [-t <mode>] [-s <factor>] [-b <number>] [-i <file>] [-m <file>] [-v]\n", name);
printf("   replicates traces recorded in <file>. Use -C prior to running this.\n");
printf("\n\
 -b --bind <number>  bind replicating process to processor number <number>\n\
 -c --convert        file to binary form, see also -o\n\
 -C --check          checks that all operations recorded in the file specied by -f will\n\
                     succeed (ie. will result in same return code).\n\
                     It takes -i and -m into account. See also -p.\n\
 -d --dont-fix       turns off fixing of missing system calls (uncomplete strace output support)\n\
 -f --file <file>    sets filename to <file>\n\
 -F --format <fmt>   specifies input format of the file.\n\
                     Options: " FORMAT_STRACE ", " FORMAT_BIN ".\n\
                     Check README for details. Default is " FORMAT_STRACE ".\n\
 -h --help           prints this message\n\
 -i --ignore <file>  sets file containing names which we should not touch during\n\
                     replaying. I.e. no syscall operation will be performed on given file.\n\
 -m --map <file>     sets containing file names mapping. When opening file,\n\
                     if there is mapping for it, it will open mapped file instead.\n\
                     See README for more information.\n\
 -o --output <file>  output filename when converting. Default: strace.bin.\n\
 -p --prepare        will prepare all files accesses recorded in file specified by -f,\n\
                     so every IO operation will return with same exit code as in original\n\
                     application. See also -i and/or -m parameters.\n\
                     Do nothing at the moment.\n\
 -P --print          prints recorded syscalls in normalized format regardless the format\n\
                     in which are the syscalls stored now\n\
 -r --replicate      will replicate every operation stored in file specified by -f\n\
 -s --scale <factor> scales delays between calls by the factor <factor>. Used with -r.\n\
 -S --stats          generate stats when processing the file. Can be combined with other\n\
                     options.\n\
 -t --timing         sets timing mode for replication. Options available:\n\
                      diff  - default mode. makes sure that gaps between calls are the same as in the original run.\n\
                      asap  - makes calls one just after another.\n\
                      exact - makes sure that calls are (approximately) done in the same time as in the original run\n\
                              (relative from start of the application)\n\
 -v --verbose be more verbose (do nothing at the moment)\n\
 -V --version prints version and exits.\n");
}

void print_version() {
	printf("ioreplay v%s, part of IOapps\n\
\n\
Copyright (C) 2010 Jiri Horky <jiri.horky@gmail.com>\n\
\n\
License GPLv2: GNU GPL version 2 <http://gnu.org/licenses/gpl.html>\n\
This is free software; see the source for copying conditions.  There is NO\n\
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n\
\n\
Written by Jiri Horky, 2010-2013\n", VERSION);
}


int main(int argc, char * * argv) {
	char format[MAX_STRING] = FORMAT_STRACE;
	char filename[MAX_STRING] = "strace.file";
	char output[MAX_STRING] = "strace.bin";
	char ignorefile[MAX_STRING] = "";
	char mapfile[MAX_STRING] = "";
	list_t * list;
	int len = 0;
	int retval;
	int action = FIX_MISSING;
	int verbose = 0;
	char c;
	int cpu;
	double scale = 1.0;
#ifdef _SC_NPROCESSORS_ONLN	 //we can determine how many processors we have
	srandom((int)time(NULL));
	cpu = random() % sysconf(_SC_NPROCESSORS_ONLN);
#else
	cpu = 0;
#endif

	gettimeofday(&global_start, NULL);

	/* Parse parameters */
	while ((c = getopt_long (argc, argv, "b:cCdf:F:hi:m:Mo:pPrs:St:vV", ioreplay_options, NULL)) != -1 ) {
		switch (c) {
			case 'b':
				cpu = atoi(optarg);
				break;
			case 'c':
				action |= ACT_CONVERT;
				break;
			case 'C':
				action |= ACT_CHECK;
				break;
			case 'd':
				action &= ~FIX_MISSING; //turn off fixing missing calls feature
				fprintf(stderr, "Turning off fix_missing...\n");
				break;
			case 'f':
				strncpy(filename, optarg, MAX_STRING);
				break;
			case 'F':
				strncpy(format, optarg, MAX_STRING);
				break;
			case 'h':
				help(basename(argv[0]));
				return 0;
				break;
			case 'i':
				strncpy(ignorefile, optarg, MAX_STRING);
				break;
			case 'm':
				strncpy(mapfile, optarg, MAX_STRING);
				break;
			case 'M':
				action |= ACT_SIMULATE;
				break;
			case 'o':
				strncpy(output, optarg, MAX_STRING);
				break;
			case 'r':
				action |= ACT_REPLICATE;
				break;
			case 'p':
				action |= ACT_PREPARE;
				break;
			case 'P':
				action |= ACT_PRINT;
				break;
			case 's':
				scale = atof(optarg);
				if (scale == 0) {
					fprintf(stderr, "Error parsing scale parameter\n");
					exit(-1);
				}
				break;
			case 'S':
				action |= ACT_STATS;
				break;
			case 't':
				if ( ! strcmp(TIME_DIFF_STR, optarg) ) {
					action = (action & ~TIME_MASK) | TIME_DIFF;
				} else if ( ! strcmp(TIME_EXACT_STR, optarg) ) {
					action = (action & ~TIME_MASK) | TIME_EXACT;
				} else if ( ! strcmp(TIME_ASAP_STR, optarg) ) {
					action = (action & ~TIME_MASK) | TIME_ASAP;
				} else {
					fprintf(stderr, "Unknown timemode specified.\n");
					exit(-1);
				}
				break;
			case 'v':
				verbose = 1;
				break;
			case 'V':
				print_version();
				return 0;
				break;
			default:
				fprintf(stderr, "Unknown parameter: %s\n", argv[optind-1]);
				return -1;
				break;
		}
	}
	
	list = (list_t *)malloc(sizeof(list_t));
	list_init(list);

	if ( !strcmp(format, FORMAT_STRACE)) {
		if ( (retval = strace_get_items(filename, list, action & ACT_STATS)) != 0) {
			DEBUGPRINTF("Error parsing file %s, exiting\n", filename);
			return retval;
		}
	} else if ( !strcmp(format, FORMAT_BIN)) {
		if ( (retval = bin_get_items(filename, list)) != 0) {
			DEBUGPRINTF("Error parsing file %s, exiting\n", filename);
			return retval;
		}
	} else {
		ERRORPRINTF("Unknown format identifier: %s\n", format);
		return -1;
	}

	len = list_length(list);
	DEBUGPRINTF("Loading done, %d items loaded.\n", len);
	char * ifilename = ignorefile;
	if (strlen(ignorefile) == 0) {
		ifilename = NULL;
	}
	char * mfilename = mapfile;
	if (strlen(mapfile) == 0) {
		mfilename = NULL;
	}

	if ( len == 0 ) {
		fprintf(stdout, "No items loaded, nothin to do --> exiting.\n");
		return 0;
	}
	if (action & ACT_PRINT) {
		DEBUGPRINTF("Listing all syscalls in normalized format...%s", "\n");
		print_items(list);
	} else if (action & ACT_CONVERT) {
		DEBUGPRINTF("Saving in binary form...%s", "\n");
		bin_save_items(output, list);
	} else if (action & ACT_SIMULATE) {
		simulate_init(ACT_SIMULATE);
		if (replicate(list, cpu, scale, action, ifilename, mfilename) != 0) {
			ERRORPRINTF("An error occurred during replicating.%s", "\n");
		}
		simulate_finish();
	} else if (action & ACT_CHECK) {
		simulate_init(ACT_CHECK);
		if (replicate(list, cpu, scale, action | ACT_SIMULATE, ifilename, mfilename) != 0) {
			ERRORPRINTF("An error occurred during replicating.%s", "\n");
		}
		simulate_check_files();
		simulate_finish();
	} else if (action & ACT_PREPARE) {
		simulate_init(ACT_PREPARE);
		if (replicate(list, cpu, scale, action, ifilename, mfilename) != 0) {
			ERRORPRINTF("An error occurred during replicating.%s", "\n");
		}
		///< @todo to change
		//simulate_prepare_files();
		simulate_finish();
	} else if (action & ACT_REPLICATE) {
		DEBUGPRINTF("Starting of replicating...%s", "\n");
		if ( ! (action & TIME_MASK) ) { //time mode not defined
			action |= TIME_DIFF; //use time diff as default
		}
		/// < @todo to change
		if (replicate(list, cpu, scale, action, ifilename, mfilename) != 0) {
			ERRORPRINTF("An error occurred during replicating.%s", "\n");
		}
	} else {
		ERRORPRINTF("No action specified!%s", "\n");
	}

	remove_items(list);
	free(list);
	return 0;
}
